Automattic/woocommerce-payments
# Use API and wordpress data to get and set gateway settings

Fixes #1590 

#### Changes proposed in this Pull Request

This adds a settings data store and fetches them from the WP REST API.

Currently supported settings are:

* Payment method list, incl. the enabled/disabled status and method order.
* "Enable WooCommerce Payments" checkbox.

## Testing instructions

### What can and cannot be tested

* Reordering, removing, and saving payment methods (but not adding - TBD in #1732).
* Enabling/disabling WCPay and saving it:

**Caveat:** Since:

1. The credit card payment method is currently the same as the main gateway (more or less).
2. The "Enable WooCommerce Payments" checkbox decides if WCPay is enabled based on the credit card gateway's `is_enabled()` return value.
3. Settings are updated in the following order: first "is WCPay enabled?", then payment methods (meaning if there's any overlap, the latter takes precedence).

... the "Enable WooCommerce Payments" checkbox will not behave as expected, even though its value is submitted during the update request and its value is retrieved from the database.

Because of the above, to actually get the "Enable WCPay" setting to update, you have to remove the CC payment method. This will change after #1705 is merged.

You can test the checkbox is actually stored by manually modifying HTTP requests, but you can also run the unit/JS tests to verify that it works as expected.

### How to test

1. Disable grouped settings (`_wcpay_feature_grouped_settings`).
1. Enable all the payment method feature flags:
  * `_wcpay_feature_giropay`
  * `_wcpay_feature_sepa`
  * `_wcpay_feature_sofort`
1. Go to http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments and enable all payment methods so you'll be able to remove and reorder them.
1. Enable grouped settings.
1. Go back to http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments.
1. Reorder/remove the payment methods as you wish, save them, refresh the page and see the order is preserved. This will be reflected on the front-end as well (e.g. during checkout).

## Follow-up issues

* Add data store support to General settings: #1727
* Add data store support to Digital wallets & express payment methods: #1728
* Add data store support to Add payment method modal: #1732
* When getting gateway order, insert payment methods after WCPay main gateway instead of replacing it: #1782
